DigitalLearningSolutions.Web/Helpers/RegistrationEmailValidator.cs

```diff
@@ -0,0 +1,71 @@
+﻿namespace DigitalLearningSolutions.Web.Helpers
+{
+    using DigitalLearningSolutions.Data.Services;
+    using DigitalLearningSolutions.Web.Extensions;
+    using DigitalLearningSolutions.Web.ViewModels.Register;
+    using Microsoft.AspNetCore.Mvc.ModelBinding;
+
+    public static class RegistrationEmailValidator
+    {
+        public static void ValidateEmailAddresses(
+            PersonalInformationViewModel model,
+            ModelStateDictionary modelState,
+            IUserService userService,
+            ICentresService? centresService = null
+        )
+        {
+            if (model.PrimaryEmail == null || !model.Centre.HasValue)
+            {
+                return;
+            }
+
+            var isRegisterAdminJourney = centresService != null;
+            var primaryEmailIsValid = !modelState.HasError(nameof(PersonalInformationViewModel.PrimaryEmail));
+            var secondaryEmailIsValid = !modelState.HasError(nameof(PersonalInformationViewModel.SecondaryEmail));
+
+            if (primaryEmailIsValid && userService.EmailIsInUse(model.PrimaryEmail))
+            {
+                modelState.AddModelError(
+                    nameof(PersonalInformationViewModel.PrimaryEmail),
+                    "A user with this primary email address is already registered; " +
+                    "if this is you, please log in at this centre via the My Account page"
+                );
+            }
+            else if (primaryEmailIsValid && isRegisterAdminJourney &&
+                     model.SecondaryEmail == null && !centresService.DoesEmailMatchCentre(
+                         model.PrimaryEmail,
+                         model.Centre.Value
+                     ))
+            {
+                modelState.AddModelError(
+                    nameof(PersonalInformationViewModel.PrimaryEmail),
+                    "This email address does not match the one held by the centre"
+                );
+            }

```

Comment:
I think I'd be in favour of nesting these else if blocks instead as it makes the code more clear.
```
if (primaryEmailIsValid){
    if (email in use)
    {
         add error
    }
    if (is register && ! matches record)
    {
        add error
    }
}
```

Especially for the secondary email, where it must be non-null to check it.

Reply:
Done. Also added a non-null check to primary email because this method gets called in the GET methods too, where the PrimaryEmail is null but valid.